### PR TITLE
review: Fix issue related to CtTypeParameterReference and getDeclaration

### PR DESCRIPTION
**Pull request by pvojtechovsky**
It is the copy of #1382 with modified solution. May be it pass all tests now?

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
Even if it pass, I do not trust that algorithm. In other cases when I develop, I know what I am doing, because I understand the context. But in this case the context is too complicated.

**Comment by surli** [comment]
> The new test only checks that it returns some CtTypeParameter. Can we somehow test that it returned the correct one?

I tried to add a new test but I'm not sure it's the kind of test you imagine. WDYT? Don't hesitate to remove my commit if you find it useless.

**Comment by surli** [comment]
@pvojtechovsky I added some more tests, tell me if it's ok for you to merge this one.

**Comment by pvojtechovsky** [comment]
I will try to have a look at Your tests today in the evening.

One idea concerning testing of correctness of type parameter references... may be we might extend code which creates Spoon model from JDT compiler AST. This would be TEST only extension, which would remember correct/expected declaration of the type parameter reference in CtElement#metadata. This extra data would allow as run a special test (e.g. on Guava library), which would check whether spoon returns correct type parameter declaration for each type parameter reference.... I really do not know if JDT compiler provides that information - I have not check that code yet ... it is just an idea how to found potential problems in code of this PR.

File: src/test/java/spoon/test/generics/GenericsTest.java [hunk 1]
```diff
@@ -1036,4 +1036,38 @@ public void testDiamondComplexGenericsRxJava() {
...
+		List<CtTypeParameter> typeParameterList = extendedPaella.getFormalCtTypeParameters();
+
+		assertEquals(1, typeParameterList.size());
+
+		CtMethod totoMethod = factory.getModel().getElements(new NameFilter<CtMethod>("toto")).get(0);
+		CtTypeReference returnTypeToto = totoMethod.getType();
+
+		CtType declaration = returnTypeToto.getDeclaration();
+
+		assertTrue(declaration.equals(typeParameterList.get(0)));
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/generics/GenericsTest.java]: I preffer asserSame instead assertEquals in this case. May be there might exist a equal type parameter, whis is not correct one.

File: src/test/java/spoon/test/generics/GenericsTest.java [hunk 2]
```diff
@@ -1036,4 +1036,38 @@ public void testDiamondComplexGenericsRxJava() {
...
+		CtMethod machinMethod = factory.getModel().getElements(new NameFilter<CtMethod>("machin")).get(0);
+		CtTypeReference returnTypeMachin = machinMethod.getType();
+		List<CtTypeParameter> formalCtTypeParameters = machinMethod.getFormalCtTypeParameters();
+
+		assertEquals(1, formalCtTypeParameters.size());
+
+		CtType declarationMachin = returnTypeMachin.getDeclaration();
+
+		assertFalse(declarationMachin.equals(typeParameterList.get(0)));
+		assertTrue(declarationMachin.equals(formalCtTypeParameters.get(0)));
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/generics/GenericsTest.java]: replace equals with == and != too. WDYT?

File: src/test/java/spoon/test/lambda/LambdaTest.java [hunk 3]
```diff
@@ -450,4 +454,17 @@ private void assertIsWellPrinted(String expected, CtLambda<?> lambda) {
...
+	@Test
+	public void testGetDeclarationOnTypeParameterFromLambda() {
+		List<CtTypeParameterReference> listCtTPR  = launcher.getModel().getElements(new TypeFilter<>(CtTypeParameterReference.class));
+
+		for (CtTypeParameterReference typeParameterReference : listCtTPR) {
+			if (!(typeParameterReference instanceof CtWildcardReference) && typeParameterReference.getDeclaration() == null) {
+				System.err.println(typeParameterReference.getSimpleName()+" from parent "+typeParameterReference.getParent(CtClass.class).getPosition()+"  has null declaration");
+				typeParameterReference.getDeclaration();
+				fail();
+			}
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/lambda/LambdaTest.java]: else { would be nice to check that returned type declaration is the correct one. The last place where it is kniwn is AST of JDT compiler.

But is just an idea. I think you can already merge this PR. The tests are fine, the code is better then before and if there is another problem hidden somewher we can fix it later.

Comment (surli) [inline src/test/java/spoon/test/lambda/LambdaTest.java]: Ok let's finish it, I need it to close another one, we can always improve the tests later :) Thanks for the comments!

**Comment by surli** [comment]
@monperrus could you review this one?

**Pull request merged**